### PR TITLE
[WIP] Add support for data write to MySQL storage table

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by sundy-li** [comment]
Almost done, could someone help review this?

File: dbms/src/Storages/StorageMySQL.cpp [hunk 1]
```diff
@@ -56,6 +64,99 @@ BlockInputStreams StorageMySQL::read(
...
+           throw;
+       }
+   }
+
+   void writeBlockData(const Block & block)
+   {
+       WriteBufferFromOwnString sqlbuf;
+       sqlbuf << "INSERT INTO  " << remote_database_name << "." << remote_table_name << " ( " << block.dumpNames() << " ) VALUES ";
+
+       auto writer = std::make_shared<BlockOutputStreamFromRowOutputStream>(std::make_shared<ValuesRowOutputStream>(sqlbuf), sample_block);
```
Comment (zhang2014) [inline dbms/src/Storages/StorageMySQL.cpp]: This should not be created for each split block write

**Comment by tobad357** [comment]
@sundy-li as a potential user it would be nice to have a config option to allow [ON DUPLICATE KEY](https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html)

**Comment by zhang2014** [comment]
I'm also interested in this task. Maybe we can work together to make it better?
I've listed several improvements that we might be able to do together : ) :
- [ ] Build write to mysql using prepared_statement
- [ ] Support `INSERT INTO  mysql() SELECT * FROM table`
- [ ] Add a new parameter at `TableFunctionMySQL` for ON DUPLICATE KEY

Look forward to your reply @sundy-li  : )

**Comment by sundy-li** [comment]
@zhang2014  Great to hear that.  Let's work it together!!!

File: dbms/src/Storages/StorageMySQL.cpp [hunk 2]
```diff
@@ -56,6 +64,99 @@ BlockInputStreams StorageMySQL::read(
...
+       writer->write(block);
+       sqlbuf << ";";
+
+       auto query = this->entry->query(sqlbuf.str());
+       query.execute();
+   }
+
+   Blocks splitBlocks(const Block & block, const size_t & max_rows) const
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMySQL.cpp]: Excessive copy when block is small enough.

File: dbms/src/Storages/StorageMySQL.cpp (hunk 1 above)
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMySQL.cpp]: Probably better way is to use FormatFactory (for looser coupling).

File: dbms/src/Storages/StorageMySQL.cpp [hunk 3]
```diff
@@ -56,6 +64,99 @@ BlockInputStreams StorageMySQL::read(
...
+       {
+           trans.rollback();
+           throw;
+       }
+   }
+
+   void writeBlockData(const Block & block)
+   {
+       WriteBufferFromOwnString sqlbuf;
+       sqlbuf << "INSERT INTO  " << remote_database_name << "." << remote_table_name << " ( " << block.dumpNames() << " ) VALUES ";
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMySQL.cpp]: backQuoteIfNeed

**Comment by alexey-milovidov** [comment]
As I remember, the most efficient way to insert data is `LOAD DATA LOCAL INFILE`, but current solution seems good enough and I don't care.

**Comment by alexey-milovidov** [comment]
> Support INSERT INTO mysql() SELECT * FROM table

I see no reason why it shouldn't work automatically.

> Add a new parameter at TableFunctionMySQL for ON DUPLICATE KEY

And REPLACE. It can be postponed for the next commit.

File: dbms/src/Storages/StorageMySQL.cpp [hunk 4]
```diff
@@ -56,6 +64,99 @@ BlockInputStreams StorageMySQL::read(
...
+        , max_batch_rows{mysql_max_rows_to_insert}
+    {
+    }
+
+   Block getHeader() const override { return sample_block; }
+
+   void write(const Block & block) override
+   {
+       auto blocks = splitBlocks(block, max_batch_rows);
+       mysqlxx::Transaction trans(entry);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMySQL.cpp]: Maybe inserting in single transaction should be optional.
Sometimes user is going to export huge amount of data and to see while it loads.

**Comment by sundy-li** [comment]
> I see no reason why it shouldn't work automatically.

Because Table functions can only be specified in the FROM clause for select usage.

> Add a new parameter at TableFunctionMySQL for ON DUPLICATE KEY

The `ON DUPLICATE KEY` and `REPLACE INTO` clauses are only used in the MySQL insert query. But from now on `table functions`  can only be  select usage.  So will it be ok to add this two config option as the engine configs?

Such as `Use ENGINE = mysql(host_port, database_name, table_name, user_name, password, on_duplicate_clause, replace_query )` . The user could create another table if he wants to have different on duplicate settings.

**Comment by zhang2014** [comment]
I created a issues[#2346] to track subsequent changes.

**Comment by alexey-milovidov** [comment]
Could you please also add something into `dbms/tests/integration`?

**Comment by sundy-li** [comment]
@alexey-milovidov    thanks for your detail comment,  I will do those improvement.

**Pull request merged**